include/core/gate.hpp

```diff
@@ -0,0 +1,277 @@
+     *
+     * @param l target line 
+     *
+
+
+     */
+        virtual void add_target(line l);
+
+        /**
+     * @brief Sets the type of the target line(s)
+     *
+     * @param t target type
+     *
+
+     */
+        virtual void set_type(const std::any& t);
+
+        /**
+     * @brief Returns the type of the target line(s)
+     *
+     * @return target type
+     *
+
+     */
+        [[nodiscard]] virtual const std::any& type() const;
+
+    private:
+        struct priv;
+        priv* const d;
+    };
+
+    struct transform_line {
+        typedef gate::line result_type;
+
+        transform_line() = default;
+
+        explicit transform_line(const std::vector<unsigned>& filter):
+            filter(&filter) {}
+
+        gate::line operator()(gate::line l) const {
+            return filter ? std::find(filter->begin(), filter->end(), l) - filter->begin() : l;
+        }
+
+    private:
+        const std::vector<unsigned>* filter = nullptr;
+    };
+    /** @endcond */
+
+    /** @cond */
+    struct filter_line {
+        filter_line() = default;
+
+        explicit filter_line(const std::vector<unsigned>& filter):
+            filter(&filter) {}
+
+        bool operator()(const gate::line& l) const {
+            return !filter || std::find(filter->begin(), filter->end(), l) != filter->end();
+        }
+
+    private:
+        const std::vector<unsigned>* filter = nullptr;
+    };
+    /** @endcond */
+
+    /**
+   * @brief Gets the control lines of a gate
+   *
+   * This function stores all control lines of a gate into a container.
+   *
+   * @section Example
+   * @code
+   * gate g = ...;
+   * std::vector<gate::line> controls;
+   * control_lines( g, std::back_inserter( controls ) );
+   * @endcode
+   *
+   * @param g      Gate
+   * @param result Iterator to store the lines as gate::line type
+   *
+   * @return The iterator after adding the lines (pointing after the end)
+   *
+
+    */
+    template<typename Iterator>
+    Iterator control_lines(const gate& g, Iterator result) {
+        for (gate::const_iterator c = g.begin_controls(); c != g.end_controls(); ++c) {
+            *result++ = *c;
+        }
+        return result;
+    }
+
+    /**
+   * @brief Gets the target lines of a gate
+   *
+   * This function stores all target lines of a gate into a container.
+   *
+   * @section Example
+   * @code
+   * gate g = ...;
+   * std::vector<gate::line> targets;
+   * target_lines( g, std::back_inserter( targets ) );
+   * @endcode
+   *
+   * @param g      Gate
+   * @param result Iterator to store the lines as gate::line type
+   *
+   * @return The iterator after adding the lines (pointing after the end)
+   *
+
+    */
+    template<typename Iterator>
+    Iterator target_lines(const gate& g, Iterator result) {
+        for (gate::const_iterator c = g.begin_targets(); c != g.end_targets(); ++c) {
+            *result++ = *c;
+        }
+        return result;
+    }
```

Comment:
these can be removed after adapting the `gate` class.
controls and targets of a gate can then easily be accessed via `g.controls` and `g.targets`.